br/pkg/restore/split/split.go

```diff
@@ -288,34 +288,88 @@ func (b *BackoffMayNotCountBackoffer) Attempt() int {
 }
 
 // GetSplitKeysOfRegions checks every input key is necessary to split region on
-// it. Returns a map from original region ID to split keys belongs to each
-// region.
+// it. Returns a map from region to split keys belongs to it.
+//
+// The key will be skipped if any of the following conditions is met:
+//   - it's the region boundary.
+//   - the key ranges consist of [prevKey, thisKey), [thisKey, nextKey) is already
+//     split.
 //
 // prerequisite:
 // - sortedKeys are sorted in ascending order.
 // - sortedRegions are continuous and sorted in ascending order by start key.
 // - sortedRegions can cover all keys in sortedKeys.
 // PaginateScanRegion should satisfy the above prerequisites.
-func GetSplitKeysOfRegions(sortedKeys [][]byte, sortedRegions []*RegionInfo, isRawKV bool) map[uint64][][]byte {
-	splitKeyMap := make(map[uint64][][]byte, len(sortedRegions))
+func GetSplitKeysOfRegions(
+	sortedKeys [][]byte,
+	sortedRegions []*RegionInfo,
+	isRawKV bool,
+) map[*RegionInfo][][]byte {
+	splitKeyMap := make(map[*RegionInfo][][]byte, len(sortedRegions))
+	pendingSplitKeyMap := make(map[*RegionInfo][][]byte, 1)
 	curKeyIndex := 0
+	splitKey := codec.EncodeBytesExt(nil, sortedKeys[curKeyIndex], isRawKV)
+	moveRegionCnt := 0
+
 	for _, region := range sortedRegions {
-		for ; curKeyIndex < len(sortedKeys); curKeyIndex += 1 {
+		for {
 			if len(sortedKeys[curKeyIndex]) == 0 {
-				continue
+				// should not happen?
+				goto nextKey
 			}
-			splitKey := codec.EncodeBytesExt(nil, sortedKeys[curKeyIndex], isRawKV)
 			// If splitKey is the boundary of the region, don't need to split on it.
 			if bytes.Equal(splitKey, region.Region.GetStartKey()) {
-				continue
+				goto nextKey
 			}
-			// If splitKey is not in a region, we should move to the next region.
+			// If splitKey is not in this region, we should move to the next region.
 			if !region.ContainsInterior(splitKey) {
+				moveRegionCnt++
 				break
 			}
-			regionID := region.Region.GetId()
-			splitKeyMap[regionID] = append(splitKeyMap[regionID], sortedKeys[curKeyIndex])
+
+			//   key1      key2    key3
+			//    |    a    |    b  |
+			// ---+----+----+----+--+-----------
+			// region1 | region2 | region3 | region4
+			//
+			// key2 don't need to be split on. Considering the original region, we can have
+			// [key1, a), [a, b), [b, key3) rather than [key1, a), [a, key2), [key2, b), [b,
+			// key3)
+
+			if moveRegionCnt > 0 {
+				clear(pendingSplitKeyMap)
+				// save key2 in pendingSplitKeyMap. When we move to key3 we can check
+				// moveRegionCnt again to know if key2 can be skipped.
+				pendingSplitKeyMap[region] = append(pendingSplitKeyMap[region], sortedKeys[curKeyIndex])
```

Comment:
seems no need to be a map, its length is <= 1 always and key is always current region